### DISCUSSION TITLE
FirebaseMessaging.onMessageOpenedApp.listen not working

### DISCUSSION THREAD

**Discussion by UmairSaqibBhutta** [firebase/flutterfire#8273] [Q&A] [answered]
I am using firebase messaging in app, everything works fine but on background notification tap, this method does not works.
I simply want to open on background notification click.

```
 FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      print('A new onMessageOpenedApp event was published!');
    
     Navigator.push(
          context,
          MaterialPageRoute(
            builder: (context) => HomePage(),
          ));
    
    });
```
when I save the code then sometimes in log `A new onMessageOpenedApp event was published` arises. But on notification click nothing works but following lines are printed:
_D/FLTFireMsgReceiver( 8637): broadcast received for message
W/FirebaseMessaging( 8637): Icon resource [https://erp.nsol.sg/img/gruopicon.png]() not found. Notification will use default icon._
And the same scenario with following method:

`  FirebaseMessaging.instance
        .getInitialMessage()
        .then((RemoteMessage? message) async {
      Navigator.push(
          context,
          MaterialPageRoute(
            builder: (context) => HomePage(),
          ));
    });`

**Comment by aymangaber66** [comment]
same issue here

**Comment by Faiz-rhm** [comment]
I have the same issue, apparently it is not working,

**Comment by psalm-byz** [comment]
any solutions here? Thanks!

**Comment by bernardoveras** [comment]
same issue here

**Comment by AlexPietri** [comment]
Same issue too with flutter version 3.3.0-0.0.pre. I'm on macOS 12.4 with AndroidStudio version 2021.1. There is a solution ?

**Comment by dsgriffin** [comment]
what's the latest working version to avoid this issue? 

`onMessageOpenedApp` recently stopped working on both iOS + Android without any changes on my side except updating deps.

**Comment by jungti1234** [comment]
same issue too! I did nothing but suddenly stopped working on both ios/android recently. what's going on?

**Comment by AlexPietri** [comment]
I removed the following statement
FirebaseMessaging.instance.setForegroundNotificationPresentationOptions and
it seems to work again on android. Unfortunately on ios I still have
configuration problems so I can't test for the moment, I will do it when I
get back from vacation.

Cordialement,
Alexandra Pietri

**Comment by jungti1234** [comment]
I solved this issue from check platform when create android channel.
create android channel only can be done on android.
if not check occur error even you don't face it.

**Comment by sahiram036** [comment]
I got it, **this issue is not about the code**. This happened because your app is in foreground when you fire the notification,
the app needs to be **terminated or in the background at this time**.
Hope this helps you...

**Reply by Anagha-vk1** [reply]
what is the code when app is in foreground

**Comment by UmairSaqibBhutta** [comment, marked as answer]
In my case, the issue was not in the code. The issue was from the server side from where I was receiving notifications in flutter app. I needed to add under given line in the server-side body of the notification.
` "click_action": "FLUTTER_NOTIFICATION_CLICK",`
and whole code is:
`  DATA='{
  "notification": {
    "body": "this is a body",
    "title": "this is a title"        
  },
  "data": {
    "att1": "value..", 
    "att2": "value..",        
    "click_action": "FLUTTER_NOTIFICATION_CLICK",
  },
  "to": "<FCM TOKEN>"
}'`

and in your manifest just add following code:
` <intent-filter>
      <action android:name="FLUTTER_NOTIFICATION_CLICK" />
      <category android:name="android.intent.category.DEFAULT" />
  </intent-filter>`

**Reply by sahiram036** [reply]
Is this code work when your app is terminated ??

**Reply by UmairSaqibBhutta** [reply]
yeah! working

**Reply by sahiram036** [reply]
Can u please share the working code !!!

**Reply by UmairSaqibBhutta** [reply]
I will share

**Reply by sahiram036** [reply]
> I will share

Kindly share your code asap...

**Reply by sakib556** [reply]
its worked thanks

**Reply by deogratiusmabima** [reply]
I face the same problem here. I've tried to use the solution you suggest but not work, could you please share your working code.

**Reply by jeffersonDSN** [reply]
it worked 
Thanks!

**Reply by UmairSaqibBhutta** [reply]
On the side, the code is very simple you can get it from the local notifications package and its backend side configuration, I have already shared in above answer.

**Reply by YawarOsman** [reply]
Not working, i also used that but same thing, when the app is terminated nothing happens when i open the app by clicking the message( i mean the app will open but onMessageOpenedApp won't get triggered).

**Comment by a-balakmet** [comment]
I hope it will be helpful.
First in method to show push notification add reader of a payload.

    void showFlutterNotification(RemoteMessage message) {
      RemoteNotification? notification = message.notification;
      AndroidNotification? android = message.notification?.android;
      if (notification != null && android != null && !kIsWeb) {
        flutterLocalNotificationsPlugin.show(
            notification.hashCode,
            notification.title,
            notification.body,
            NotificationDetails(
              android: AndroidNotificationDetails(
                channel.id,
                channel.name,
                channelDescription: channel.description,
                // TODO add a proper drawable resource to android, for now using
                // one that already exists in example app.
                icon: 'launch_background',
              ),
            ),
            // Add reading of a payload based on the contents of notification
            // In my case:
            payload: notification.data['params']);
      }
    }

Then somewhere in the application for example in initState method try to call the payload.


    NotificationAppLaunchDetails? notificationAppLaunchDetails = await flutterLocalNotificationsPlugin.getNotificationAppLaunchDetails();
        if (notificationAppLaunchDetails != null && notificationAppLaunchDetails.didNotificationLaunchApp) {
          var payload = notificationAppLaunchDetails.notificationResponse?.payload;
          if (payload != null && payload.isNotEmpty) {
            // Your code to handle the payload for routing.
          }
        }

Be aware that flutterLocalNotificationsPlugin is properly initialized in main().

**Reply by Silfalion** [reply]
This is the solution that also worked for me, however what I found is that it's problematic when using grouped notifications. Since you have to use flutter local notifications to display them rather than leaving it to Firebase, the payload won't be able to contain a notification key, but that's not really an issue. The issue is that firebase not creating the notifications, it won't be able to trigger the onMessageOpenedApp function

**Reply by YawarOsman** [reply]
worked for me

**Comment by vigneshaskan** [comment]
I'm also facing the remote message navigation issue.
In android the redirection is working fine when i click on push notification.
But in IOS it just opening app and the redirection is not happening.

FirebaseMessaging.onMessage.listen(
      (RemoteMessage message) {
      print('test');
      });
 FirebaseMessaging.onMessageOpenedApp.listen(
      (RemoteMessage message) {
      print('test');
      });
      
Both not working on IOS

**Reply by UmairSaqibBhutta** [reply]
Use this listener :
```
FirebaseMessaging.instance
        .getInitialMessage()
        .then((RemoteMessage? message) {
// you navigation code here
}
```

**Comment by Abhaykumarbhumihar** [comment]
import 'package:firebase_messaging/firebase_messaging.dart';

class MessagingService {
  final FirebaseMessaging _fcm = FirebaseMessaging.instance;

  Future<void> init() async {
    _fcm.requestPermission();

    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      print("onMessage:onMessage  $message");

    });

    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      print("onMessageOpenedApp: $message");
      _handleNotificationClick(message);
    });
  }

  void _handleNotificationClick(RemoteMessage message) {
    print("_handleNotificationClick _handleNotificationClick _handleNotificationClick");
    // Extract data from the message
    final notificationData = message.data;

    // Perform the desired action
    if (notificationData.containsKey('screen')) {


    } else {
      // Handle other types of notifications
    }
  }
}

 FirebaseMessaging.onMessageOpenedApp.listen is not working when app is open, and on tap of notification bar

**Comment by mansur61** [comment]
Firabase entegrasyonlarınız düzgün yüklenmiş mıdır?. Mainde fonksiyonunda
ilgili initial entegre bilgileri bulunmakta mıdır?

16 Oca 2026 Cum 15:23 tarihinde Yawar Othman ***@***.***>
şunu yazdı:

**Reply by YawarOsman** [reply]
Firebase works perfect, ios (background, foreground, terminated) states work perfectly, android (background, foreground) works perfectly, but when the app is terminated it's not working, i get the notification and when i press it it opens the app but not doing the redirection

**Comment by mansur61** [comment]
Merhabalar,
İlgili paket yönlendirme hizmeti sunmamaktadır.
Farklı bir versiyonda ilgili talebinizi gerçekleştirmek için notlarımı
aldım

Saygılarımla

16 Oca 2026 Cum 15:57 tarihinde Yawar Othman ***@***.***>
şunu yazdı:

**Reply by YawarOsman** [reply]
This one just solved my issue, thanks


https://github.com/firebase/flutterfire/discussions/8273?notification_referrer_id=NT_kwHOBBG_-doAJ1JlcG9zaXRvcnk7MjAyNDI5NzAzO0Rpc2N1c3Npb247MzkzNTk2Nw#discussioncomment-5174855

**Comment by mansur61** [comment]
Veya yönlendirmeden kastınız nedir? Ekran görüntüsü paylaşabilir misiniz

17 Oca 2026 Cmt 01:05 tarihinde Mansuremin Kaya ***@***.***>
şunu yazdı:

**Comment by mansur61** [comment]
Başarılar dilerim.


18 Oca 2026 Paz 18:29 tarihinde Yawar Othman ***@***.***>
şunu yazdı:

**Comment by mansur61** [comment]
Umarım paket kullanımını beğenmişsinizdir
Sağlıklı günler dilerim

18 Oca 2026 Paz 19:08 tarihinde Mansuremin Kaya ***@***.***>
şunu yazdı:
